File: sdk/src/main/java/com/libreshockwave/chunks/CastChunk.java
package com.libreshockwave.chunks;

import com.libreshockwave.DirectorFile;
import com.libreshockwave.format.ChunkType;
import com.libreshockwave.io.BinaryReader;
import com.libreshockwave.vm.LingoVM;

import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.List;

/**
 * Cast chunk (CAS*).
 * Contains array of cast member IDs for a cast library.
 */
public record CastChunk(
    DirectorFile file,
    int id,
    List<Integer> memberIds
) implements Chunk {

    @Override
    public ChunkType type() {
        return ChunkType.CASp;
    }

    public int memberCount() {
        return memberIds.size();
    }

    public static CastChunk read(DirectorFile vm,  BinaryReader reader, int id, int version) {
        // CAS* chunk is always big endian regardless of file byte order
        reader.setOrder(ByteOrder.BIG_ENDIAN);

        List<Integer> memberIds = new ArrayList<>();

        while (reader.bytesLeft() >= 4) {
            memberIds.add(reader.readI32());
        }

        return new CastChunk(vm, id, memberIds);
    }
}

File: cpp/src/chunks/CastChunk.cpp
#include "libreshockwave/chunks/CastChunk.hpp"

#include "libreshockwave/io/BinaryReader.hpp"

namespace libreshockwave::chunks {

CastChunk::CastChunk(const DirectorFile* file, id::ChunkId id, std::vector<int> memberIds)
    : file_(file), id_(id), memberIds_(std::move(memberIds)) {}

const DirectorFile* CastChunk::file() const { return file_; }
format::ChunkType CastChunk::type() const { return format::ChunkType::CASp; }
id::ChunkId CastChunk::id() const { return id_; }
const std::vector<int>& CastChunk::memberIds() const { return memberIds_; }
int CastChunk::memberCount() const { return static_cast<int>(memberIds_.size()); }

CastChunk CastChunk::read(const DirectorFile* file, io::BinaryReader& reader, id::ChunkId id, int version) {
    (void)version;
    reader.setOrder(io::ByteOrder::BigEndian);
    std::vector<int> memberIds;
    while (reader.bytesLeft() >= 4) {
        memberIds.push_back(reader.readI32());
    }
    return CastChunk(file, id, std::move(memberIds));
}

} // namespace libreshockwave::chunks
